brozzler/frontier.py

```diff
@@ -30,6 +33,57 @@ class UnexpectedDbResult(Exception):
     pass
 
 
+def filter_claimable_site_ids(
+    active_sites: List[Dict], max_sites_to_claim=1
+) -> List[str]:
+    job_counts = {}
+    claimable_sites = []
+    now = datetime.datetime.now(datetime.timezone.utc)
+
+    for site in active_sites:
+        is_claimable = False
+
+        # If site not claimed and not disclaimed within last 20 seconds
+        if not site["claimed"] and site.get("last_disclaimed", 0) <= (
+            now - datetime.timedelta(seconds=20)
+        ):
+            is_claimable = True
+
+        # or site has been disclaimed more than an hour ago
+        if "last_claimed" in site and site["last_claimed"] <= (
+            now - datetime.timedelta(hours=1)
+        ):
+            is_claimable = True
+
+        # Count number of claimed sites per job_id (optional field)
+        if site["claimed"] and "max_claimed_sites" in site and "job_id" in site:
+            job_id = site["job_id"]
+            job_counts[job_id] = job_counts.get(job_id, 0) + 1
+
+        if is_claimable:
+            claimable_sites.append(site)
+
+    site_ids_to_claim = []
+    # gather sites that are under the max without going over
+    for site in claimable_sites:
+        if (
+            "max_claimed_sites" in site
+            and "job_id" in site
+            and job_counts.get(site["job_id"], 0) < site["max_claimed_sites"]
+        ):
+            site_ids_to_claim.append(site["id"])
```

Comment:
we could do this before the `if` (and include only the job_counts update after `if`, and skip the second `if` block)

Reply:
These 2 if clauses are also effectively a large `or`. Merging them in this way fails some test cases where we exceed max claimed sites when it is provided